funsor/numpy.py

```diff
@@ -299,3 +299,151 @@ def materialize(x):
         subs.append((name, arange(name, domain.dtype)))
     subs = tuple(subs)
     return substitute(x, subs)
+
+
+################################################################################
+# Register Ops
+################################################################################
+
+
+@ops.sigmoid.register(np.ndarray)
+def _sigmoid(x):
+    try:
+        from scipy.special import expit
+        return expit(x)
+    except:
+        return 1 / (1 + np.exp(-x))
+
+
+@ops.sqrt.register(np.ndarray)
+def _sqrt(x):
+    return np.sqrt(x)
+
+
+@ops.exp.register(np.ndarray)
+def _exp(x):
+    return np.exp(x)
+
+
+@ops.log.register(np.ndarray)
+def _log(x):
+    return np.log(x)
+
+
+@ops.log1p.register(np.ndarray)
+def _log1p(x):
+    return np.log1p(x)
+
+
+@ops.min.register(np.ndarray, np.ndarray)
+def _min(x, y):
+    return np.minimum(x, y)
+
+
+@ops.min.register((int, float), np.ndarray)
+def _min(x, y):
+    return np.clip(y, a_max=x)
+
+
+@ops.min.register(np.ndarray, (int, float))
+def _min(x, y):
+    return np.clip(x, a_max=y)
+
+
+@ops.max.register(np.ndarray, np.ndarray)
+def _max(x, y):
+    return np.maximum(x, y)
+
+
+@ops.max.register((int, float), np.ndarray)
+def _max(x, y):
+    return np.clip(y, a_min=x)
+
+
+@ops.max.register(np.ndarray, (int, float))
+def _max(x, y):
+    return np.clip(x, a_min=y)
+
+
+@ops.reciprocal.register(np.ndarray)
+def _reciprocal(x):
+    result = np.clip(np.reciprocal(x), a_max=np.finfo(x.dtype).max)
+    return result
+
+
+@ops.safesub.register((int, float), np.ndarray)
+def _safesub(x, y):
+    try:
```

Comment:
1. It would be best to avoid slow try-except flow in every operation; maybe we can cache these.
2. Let's at least limit the scope of the try-except
```py
try:
    finfo = np.finfo(y.dtype)
except ValueError:
    finfo = np.iinfo(y.dtype)
return x + np.clip(-y, max=finfo.max)
```